### ISSUE TITLE
the widely not understood ORA-03106 can be reproduced w/ cx_Oracle's stmt cache feature ( as well as ORA-00932)

### ISSUE THREAD

**Issue by zzzeek** [oracle/python-cx_Oracle#519]
Hey just a heads up, I've had this issue for likely ten years and i've finally figured out a large part of it.  If you google for "ORA-03106: fatal two-task communication protocol error", even [Burleson](http://www.dba-oracle.com/sf_ora_03106_fatal_two_task_communication_protocol_error.htm) has almost nothing to say on it (though what he wrote there gave me a clue).

short answer, if you drop a table and create it with different datatypes, then run the same kind of SELECT as before on it and throw in some cursor.var(), you can get ORA-03106 or with some modification an ORA-00932.  both are related to the statement cache so if you set cx_Oracle.stmtcachesize = 0, both conditions resolved.  it might be good to note this at https://cx-oracle.readthedocs.io/en/latest/user_guide/tuning.html#stmtcache or maybe even allow for an up front "stmtcachesize" connection parameter, as for our test suite I need to set this to zero if I am trying to re-use connections across many tests.

Example below on cx_Oracle 8.1.0 I can use to reproduce both ORA-03106 and ORA-00932.

```python
import cx_Oracle

conn = cx_Oracle.connect(user="scott", password="tiger", dsn="oracle18c")


def run_select():

    cursor = conn.cursor()
    cursor.execute("SELECT x FROM t")
    cursor.fetchall()
    cursor.close()
    conn.rollback()


def run_clob_select_for_00932():
    cursor = conn.cursor()
    cursor.execute("SELECT x FROM t")
    cursor.close()
    conn.rollback()


def run_clob_select_for_03106():
    def handler(cursor, name, defaultType, size, precision, scale):
        if name == "X":
            return cursor.var(
                cx_Oracle.STRING, 255, arraysize=cursor.arraysize
            )

    cursor = conn.cursor()
    cursor.outputtypehandler = handler
    cursor.execute("SELECT x FROM t")
    cursor.fetchall()
    cursor.close()
    conn.rollback()


def create_varchar_table():
    cursor = conn.cursor()

    try:
        cursor.execute("DROP TABLE t")
    except:
        pass
    cursor.execute(
        """
    CREATE TABLE t (
        x VARCHAR2(255 CHAR)
    )
    """
    )
    cursor.execute("INSERT INTO t (x) VALUES ('some value')")
    cursor.close()
    conn.commit()


def create_clob_table():
    cursor = conn.cursor()

    try:
        cursor.execute("DROP TABLE t")
    except:
        pass
    cursor.execute(
        """
    CREATE TABLE t (
        x CLOB
    )
    """
    )
    cursor.execute("INSERT INTO t (x) VALUES ('some value')")
    cursor.close()
    conn.commit()


# turn off the cache, no error.  leave cache on,
# the two errors can occur.
# conn.stmtcachesize = 0

create_clob_table()

# run this w/ statement caching to reproduce ORA-03106
run_clob_select_for_03106()

# alternatively, this one will do 00932
# run_clob_select_for_00932()

create_varchar_table()
run_select()

```

**Comment by anthony-tuininga** [comment]
Yes, this is a "known" issue and the "solution" is to disable the statement cache. The bug number for those wanting to request this via Oracle Support is 25184056. A patch is available internally for this bug which will become available as part of an official release at some point -- exactly when I cannot say! Your suggestion to add a parameter to `connect()` that allows the statement cache size to be set is a reasonable one. I'll ask internally and if there are no objections I'll create a patch for that. Thanks for reporting this with such a nice reproducer!

**Comment by zzzeek** [comment]
has cx_Oracle had this parameter for many years and I just didn't know about it?

**Comment by anthony-tuininga** [comment]
Not at all! You have to set the attribute after you create the connection right now. Your suggestion to add the parameter seems reasonable to me. If I didn't make that clear, my apologies!

**Comment by zzzeek** [comment]
I meant, was the attribute on the connection available.  like back in cx_oracle 5, 4, etc.

**Comment by anthony-tuininga** [comment]
Ah! Connections had the attribute `stmtcachesize` since cx_Oracle 4.3.2 (August 2007). It was added to session pools in version 6 (April 2017). So yeah, its been around for a while!

**Comment by zzzeek** [comment]
so I should also add, that if there  were a way to **reset** this cache, that would serve my purposes also, as the issue is that after DDL is emitted, the caches are out of date.   I have the same issue with the [asyncpg](https://github.com/MagicStack/asyncpg) driver where I make use of the [reload_schema_state](https://magicstack.github.io/asyncpg/current/api/index.html?highlight=reload_schema_state#asyncpg.connection.Connection.reload_schema_state) method after DDL has changed things around.    that way I could still be testing SQLAlchemy against cx_Oracle in its normal mode of operation.

**Comment by anthony-tuininga** [comment]
You can change the value of `stmtcachesize` to 0 and then change it back again to whatever value you would like. This will flush the cache completely and then start building it again afterwards. I believe that would do what you want it to do.

**Comment by zzzeek** [comment]
ah interesting!   more things to add to the docs :)

**Comment by cjbj** [comment]
If resetting the cache is usable, and with 25184056 coming, and with ODPI-C already protecting against some cases, and if this is only for test scenarios, do we really want to add yet-another-option?  Note the statement cache size can also be set in [oraacess.xml](https://docs.oracle.com/en/database/oracle/oracle-database/21/lnoci/performance-topics.html#GUID-C69F99D2-2EAC-4A5E-BD7D-7266765C45C2).

How about adding a 'drop' mode to cursor.close(), similar to the request in https://github.com/oracle/node-oracledb/issues/182?

**Comment by cjbj** [comment]
@zzzeek The python-oracledb Thin mode handles this scenario. A partial fix for 25184056 in the Oracle Client libraries used by cx_Oracle and python-oracledb Thick mode has already been merged to a future Oracle version - we're pushing to have a more complete fix.  Python-oracledb has a per-statement drop-from-cache feature, see https://python-oracledb.readthedocs.io/en/latest/api_manual/cursor.html#Cursor.prepare
